cpp/src_prims/sparse/batched/csr.h

```diff
@@ -0,0 +1,721 @@
+    auto deallocate_values = [allocator, batch_size, nnz, stream](T* A) {
+      allocator->deallocate(A, batch_size * nnz * sizeof(T), stream);
+    };
+    auto deallocate_col = [allocator, nnz, stream](int* A) {
+      allocator->deallocate(A, sizeof(int) * nnz, stream);
+    };
+    auto deallocate_row = [allocator, shape, stream](int* A) {
+      allocator->deallocate(A, sizeof(int) * (shape.first + 1), stream);
+    };
+
+    // When the shared pointers go to 0, the memory is deallocated
+    m_values = std::shared_ptr<T>(values, deallocate_values);
+    m_col_index = std::shared_ptr<int>(col_index, deallocate_col);
+    m_row_index = std::shared_ptr<int>(row_index, deallocate_row);
+  }
+
+  /**
+   * @brief Construct from a dense batched matrix and its mask
+   * 
+   * @param[in]  dense  Dense batched matrix
+   * @param[in]  mask   Col-major host device matrix containing a mask of the
+   *                    non-zero values common to all matrices in the batch.
+   *                    Note: the point of using a mask is that some values
+   *                    might be zero in a few matrices but not generally in
+   *                    the batch so we shouldn't rely on a single matrix to
+   *                    get the mask
+   * @return Batched CSR matrix
+   */
+  static CSR<T> from_dense(const LinAlg::Batched::Matrix<T>& dense,
+                           const std::vector<bool>& mask,
+                           cusolverSpHandle_t cusolverSpHandle) {
+    std::pair<int, int> shape = dense.shape();
+
+    // Create the index arrays from the mask
+    std::vector<int> h_col_index;
+    std::vector<int> h_row_index = std::vector<int>(shape.first + 1);
+    int nnz = 0;
+    for (int i = 0; i < shape.first; i++) {
+      h_row_index[i] = nnz;
+      for (int j = 0; j < shape.second; j++) {
+        if (mask[j * shape.first + i]) {
+          h_col_index.push_back(j);
+          nnz++;
+        }
+      }
+    }
+    h_row_index[shape.first] = nnz;
+
+    CSR<T> out = CSR<T>(shape.first, shape.second, nnz, dense.batches(),
+                        dense.cublasHandle(), cusolverSpHandle,
+                        dense.allocator(), dense.stream());
+
+    // Copy the host index arrays to the device
+    MLCommon::copy(out.get_col_index(), h_col_index.data(), nnz, out.stream());
+    MLCommon::copy(out.get_row_index(), h_row_index.data(), shape.first + 1,
+                   out.stream());
+
+    // Copy the data from the dense matrix to its sparse representation
+    constexpr int TPB = 256;
+    dense_to_csr_kernel<<<MLCommon::ceildiv<int>(out.batches(), TPB), TPB, 0,
+                          out.stream()>>>(
+      dense.raw_data(), out.get_col_index(), out.get_row_index(),
+      out.get_values(), out.batches(), shape.first, shape.second, nnz);
+    CUDA_CHECK(cudaPeekAtLastError());
+
+    return out;
+  }
+
+  /**
+   * @brief Construct a dense batched matrix
+   * 
```

Comment:
better to add a note about the asynchronicity of this method? (meaning this function could return before the output matrix contents have been populated by the cuda kernel)